packages/airnode-deployer/src/infrastructure/gcp.ts

```diff
@@ -1,22 +1,181 @@
 import { GcpCloudProvider } from '@api3/airnode-node';
+import { go } from '@api3/promise-utils';
 import { Storage } from '@google-cloud/storage';
+import {
+  BUCKET_NAME_REGEX,
+  Directory,
+  FileSystemType,
+  generateBucketName,
+  translatePathsToDirectoryStructure,
+} from '../utils/infrastructure';
 import * as logger from '../utils/logger';
 
-export async function stateExists(bucketName: string, cloudProvider: GcpCloudProvider) {
-  logger.debug('Checking Terraform state existence in GCP');
-  const { projectId } = cloudProvider;
-  const storage = new Storage({ projectId });
-  const bucket = storage.bucket(bucketName);
+const initializeGcsService = (cloudProvider: GcpCloudProvider) => new Storage({ projectId: cloudProvider.projectId });
 
-  return (await bucket.exists())[0] as boolean;
-}
+const initializeGcsBucket = (cloudProvider: GcpCloudProvider, bucketName: string) =>
+  initializeGcsService(cloudProvider).bucket(bucketName);
 
-export async function removeState(bucketName: string, cloudProvider: GcpCloudProvider) {
-  logger.debug('Removing Terraform state from GCP');
-  const { projectId } = cloudProvider;
-  const storage = new Storage({ projectId });
-  const bucket = storage.bucket(bucketName);
+export const getAirnodeBucket = async (cloudProvider: GcpCloudProvider) => {
+  const storage = initializeGcsService(cloudProvider);
 
-  await bucket.deleteFiles({ force: true, versions: true });
-  await bucket.delete();
-}
+  logger.debug('Listing GCS buckets');
+  const goBuckets = await go(() => storage.getBuckets());
+  if (!goBuckets.success) {
+    throw new Error(`Failed to list GCS buckets: ${goBuckets.error}`);
+  }
+
+  const [buckets] = goBuckets.data;
+  const airnodeBuckets = buckets.filter((bucket) => bucket.name.match(BUCKET_NAME_REGEX));
+  if (airnodeBuckets.length > 1) {
+    throw new Error(`Multiple Airnode buckets found, stopping. Buckets: ${JSON.stringify(airnodeBuckets)}`);
+  }
+
+  return airnodeBuckets[0]?.name || null;
+};
+
+export const createAirnodeBucket = async (cloudProvider: GcpCloudProvider) => {
+  const storage = initializeGcsService(cloudProvider);
+  const bucketName = generateBucketName();
+
+  logger.debug(`Creating GCS bucket '${bucketName}'`);
+  const goCreate = await go(() => storage.createBucket(bucketName));
+  if (!goCreate.success) {
+    throw new Error(`Failed to create an GCS bucket: ${goCreate.error}`);
+  }
+
+  const [bucket] = goCreate.data;
+
+  logger.debug(`Setting uniform bucket-level access for GCS bucket '${bucketName}'`);
+  const goMetadata = await go(() =>
+    bucket.setMetadata({
+      iamConfiguration: { uniformBucketLevelAccess: { enabled: true }, publicAccessPrevention: 'enforced' },
+    })
+  );
+  if (!goMetadata.success) {
+    throw new Error(`Failed to setup a uniform bucket-level access for bucket '${bucketName}': ${goMetadata.error}`);
+  }
+
+  const policy = {
+    bindings: [
+      {
+        role: 'roles/storage.legacyBucketReader',
```

Comment:
Can we document why we need this? I am struggling to understand why many of the things like these are needed... (Same goes for the AWS logic).

Basically anything that has to be set in cloud is hard to reason about, whether it must be set and why. I fear this will be painful to modify in a few months advance.